sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala

```diff
@@ -1,66 +1,121 @@
 package sigmastate.interpreter
 
-import sigmastate.{ProofTree, SigSerializer, UncheckedConjecture, UncheckedLeaf, UncheckedSigmaTree}
+import sigmastate.{ProofTree, SigSerializer, SigmaConjecture, SigmaProofOfKnowledgeLeaf, UncheckedConjecture, UncheckedLeaf, UncheckedSigmaTree}
 import sigmastate.Values.{ErgoTree, SigmaBoolean}
+import sigmastate.basics.DLogProtocol.{DLogInteractiveProver, ProveDlog}
+import sigmastate.basics.{DiffieHellmanTupleInteractiveProver, ProveDHTuple}
 import sigmastate.basics.VerifierMessage.Challenge
 
 
 trait ProverUtils extends Interpreter {
 
+  /**
+    * Generate commitments for a given ergoTree (mixed-tree) and public keys.
+    *
+    * First, the given tree is to be reduced to crypto-tree (sigma-tree) by using context provided.
+    */
+  def generateCommitmentsFor(ergoTree: ErgoTree,
+                             context: CTX,
+                             generateFor: Seq[SigmaBoolean]): HintsBag = {
+    val reducedTree = fullReduction(ergoTree, context, Interpreter.emptyEnv)._1
+    generateCommitmentsFor(reducedTree, generateFor)
+  }
+
+  /**
+    * A method which is is generating commitments for all the public keys provided.
+    *
+    * Currently only ProveDlog and ProveDiffieHellman are supported.
+    *
+    * @param sigmaTree - crypto-tree
+    * @param generateFor - public keys
+    * @return
+    */
+  def generateCommitmentsFor(sigmaTree: SigmaBoolean,
+                             generateFor: Seq[SigmaBoolean]): HintsBag = {
+
+    def traverseNode(sb: SigmaBoolean,
+                     bag: HintsBag,
+                     position: String): HintsBag = {
+      sb match {
+        case sc: SigmaConjecture =>
+          sc.children.zipWithIndex.foldLeft(bag) { case (b, (child, idx)) =>
+            traverseNode(child, b, position + "-" + idx)
+          }
+        case leaf: SigmaProofOfKnowledgeLeaf[_, _] =>
+          if (generateFor.contains(leaf)) {
+            val (r, a) = leaf match {
+              case _: ProveDlog =>
+                DLogInteractiveProver.firstMessage()
+              case pdh: ProveDHTuple =>
+                DiffieHellmanTupleInteractiveProver.firstMessage(pdh)
+              case _ => ???
+            }
+            val hints = Seq(OwnCommitment(leaf, r, a, position), RealCommitment(leaf, a, position))
+            bag.addHints(hints: _*)
+          } else {
+            bag
+          }
+      }
+    }
+
+    traverseNode(sigmaTree, HintsBag.empty, position = Hint.CryptoTreePrefix)
+  }
```

Comment:
Please add test-vector-like tests for this method, especially boundary cases and negative case, which are not covered now. 
This is to catch regressions, because applications will rely on exact behavior of this code.

Reply:
test done, see ProverSpecification

Reply:
DHT case is not covered
<img width="991" alt="image" src="https://user-images.githubusercontent.com/654077/93190626-42e37700-f74c-11ea-80b6-2ace587d41a4.png">

Reply:
done